README.md
<h1 align="center">Easy Editor</h1>
<div align="center">
<img src="https://img.shields.io/npm/v/@camol/easy-editor?style=flat-square"/>
<img alt="GitHub" src="https://img.shields.io/github/license/kanweiwei/easy-editor">
</div>
<div align="center">A rich text editor based on slate </div>

![easy-editor](./source/images/easy-editor-demo.gif)

[Demo](https://kanweiwei.github.io/easy-editor/example/)

## Quick Start

### npm

```bash
npm install @camol/easy-editor --save
```

### yarn

```bash
yarn add @camol/easy-editor -S
```

### Examples

```bash
git clone https://github.com/kanweiwei/easy-editor.git
cd easy-editor
yarn && yarn build && yarn example
```

### with React

```typescript
import * as React from "react";
import ReactDom from "react-dom";
import EasyEditor from "@camol/easy-editor";

class Editor extends React.Component {
  html = "";

  handleChange = (v: any) => {
    console.log("change=>>>", v);
    if (this.editorRef.current) {
      // value to html
      console.log(this.editorRef.current.convertor.serialize(v.change.value));
    }
  };

  render() {
    return <EasyEditor value={"<p>123</p>"} onChange={this.handleChange} />;
  }
}

ReactDom.render(<Editor />, document.getElementById("root"));
```

### Custom toolbar

For convenience, some tools have been built in the toolbar.
tool| instructions|
-|:-:|
fontSize| 字号|
fontColor| 字体颜色|
bold| 加粗|
italic| 斜体|
u| 下划线|
sup| 上标 |
sub | 下标 |
left|文字居左|
right|文字居右|
center|文字居中|
justify|两端对齐|
image| 插入图片|
video | 插入音频（mp4、webm）|
pdf | 插入 pdf|
math | 插入 latex 数学公式 |

#### How to add a custom tool ?

```typescript
import * as React from "react";
import ReactDom from "react-dom";
import EasyEditor from "@camol/easy-editor";

// audio tool
class AudioControl extends React.Component {
  inputRef = React.createRef();

  handleClick = () => {
    if (inputRef.current) {
      inputRef.current.click();
    }
  };

  handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const file = e.target.files[0];
    e.target.value = "";
    if (file) {
      if (this.props.beforeUpload) {
        let url = await this.props.beforeUpload(file);
        if (url) {
          let change = this.props.change.focus().insertInline({
            object: "inline",
            type: "audio",
            isVoid: true,
            data: {
              src: url,
            },
          });
          this.props.update(change);
        }
      }
    }
  };

  render() {
    return (
      <span onMouseDown={this.handleClick}>
        <span className="tool-insert-video" />
        <input
          type="file"
          style={{ width: 0, height: 0, opacity: 0 }}
          ref={this.inputRef}
          onChange={this.handleChange}
        />
      </span>
    );
  }
}

class Editor extends React.Component {
  html = "";

  handleChange = (v: any) => {
    console.log("change=>>>", v);
    if (this.editorRef.current) {
      // value to html
      console.log(this.editorRef.current.convertor.serialize(v.change.value));
    }
  };

  render() {
    return (
      <EasyEditor
        value={"<p>123</p>"}
        onChange={this.handleChange}
        controls={[
          ["bold", "u", "image"],
          [
            {
              type: "audio",
              component: AudioControl,
            },
          ],
        ]}
      />
    );
  }
}
```

---

### Custom Node Plugin

```typescript
import * as React from "react";
import { AST } from "parse5";

const plugin = {
  type: "node", // node, mark
  object: "inline",
  nodeType: "super-audio",
  importer(el: AST.Default.Element, next: Function): any {
    if (el.tagName.toLowerCase() === "audio") {
      return {
        object: "inline", // block、inline,
        type: "super-audio",
        isVoid: true,
        data: {
          src: el?.attrs?.find((n) => n.name === "src")?.value,
        },
        nodes: next(el.childNodes),
      };
    }
  },
  exporter(node: any, children: any): any {
    let { className, src } = node.data.toJS();
    return <audio src={src} className={className}></audio>;
  },
  render(
    editor: any,
    props: { attributes: any; children: any; node: any; isSelected: any }
  ): any {
    // @ts-ignore
    const { attributes, children, node, isSelected } = props;
    const src = node.data.get("src");
    return (
      <span {...attributes}>
        <audio src={src} controls>
          {children}
        </audio>
      </span>
    );
  },
};

export default plugin;
```

---

### License

Apache 2.0


CHANGELOG.md
# [0.4.0](https://github.com/kanweiwei/easy-editor/compare/v0.3.13...v0.4.0) (2020-09-07)


### Bug Fixes

* image url ([851e8e9](https://github.com/kanweiwei/easy-editor/commit/851e8e9d5a3291c63bdafbadcca672c343cdb36c))


### Features

* support adjust font color ([05a4179](https://github.com/kanweiwei/easy-editor/commit/05a417950efc2fbebf18e2c173b9fc9737aaba42))
* support adjust font size ([9b91b50](https://github.com/kanweiwei/easy-editor/commit/9b91b50b9a6efda784c7401c6256031d0ec0b0c6))



## [0.3.13](https://github.com/kanweiwei/easy-editor/compare/v0.3.12...v0.3.13) (2020-09-02)



## [0.3.12](https://github.com/kanweiwei/easy-editor/compare/v0.3.11...v0.3.12) (2020-09-02)


### Bug Fixes

* fix style when pdf is readOnly; adjust pdf height of pdfView ([7a56bb8](https://github.com/kanweiwei/easy-editor/commit/7a56bb871e66508ba75e5055de4189c8183aeee2))



## [0.3.11](https://github.com/kanweiwei/easy-editor/compare/v0.3.10...v0.3.11) (2020-09-01)


### Bug Fixes

* pdf name ([e17202d](https://github.com/kanweiwei/easy-editor/commit/e17202d41894b1649d94232b76e4cfd2c4d63334))



## [0.3.10](https://github.com/kanweiwei/easy-editor/compare/v0.3.9...v0.3.10) (2020-09-01)


### Bug Fixes

* pdf url ([9ddfc83](https://github.com/kanweiwei/easy-editor/commit/9ddfc8344ba64173478cfe5e4cd100effc269344))



## [0.3.9](https://github.com/kanweiwei/easy-editor/compare/v0.3.8...v0.3.9) (2020-08-28)


### Features

* add video renderNode ([262d186](https://github.com/kanweiwei/easy-editor/commit/262d186f0855499877cd08d90bf061108b1da3e6))



## [0.3.8](https://github.com/kanweiwei/easy-editor/compare/v0.3.7...v0.3.8) (2020-08-24)


### Bug Fixes

* fix playback issue ([80201bf](https://github.com/kanweiwei/easy-editor/commit/80201bf3ec9675cec76ebf77b647386380a05f93))



## [0.3.7](https://github.com/kanweiwei/easy-editor/compare/v0.3.6...v0.3.7) (2020-08-14)


### Bug Fixes

* add serialize for '<a></a>' ([256f3ff](https://github.com/kanweiwei/easy-editor/commit/256f3fff50aa2f230d91e40bf4642783fad0add3))



## [0.3.6](https://github.com/kanweiwei/easy-editor/compare/v0.3.5...v0.3.6) (2020-08-13)



## [0.3.5](https://github.com/kanweiwei/easy-editor/compare/v0.3.4...v0.3.5) (2020-08-11)


### Bug Fixes

* image-edit-toobar style ([0f13221](https://github.com/kanweiwei/easy-editor/commit/0f13221f0905bfc7ff3f317784cb8f8c18d37daf))



## [0.3.4](https://github.com/kanweiwei/easy-editor/compare/v0.3.3...v0.3.4) (2020-08-11)



## [0.3.3](https://github.com/kanweiwei/easy-editor/compare/v0.3.2...v0.3.3) (2020-08-11)



## [0.3.2](https://github.com/kanweiwei/easy-editor/compare/v0.3.1...v0.3.2) (2020-08-11)


### Features

* eslint ([396f58e](https://github.com/kanweiwei/easy-editor/commit/396f58e2149ccc4cfe31cd7de4489aac0416c9b9))



## [0.3.1](https://github.com/kanweiwei/easy-editor/compare/v0.3.0...v0.3.1) (2020-08-10)



# [0.3.0](https://github.com/kanweiwei/easy-editor/compare/v0.2.4...v0.3.0) (2020-08-04)



## [0.2.4](https://github.com/kanweiwei/easy-editor/compare/v0.2.3...v0.2.4) (2020-07-27)


### Bug Fixes

* remove repeat code ([8d4bf2e](https://github.com/kanweiwei/easy-editor/commit/8d4bf2e98c115612f4bad30684e097e606a079c2))



## [0.2.3](https://github.com/kanweiwei/easy-editor/compare/v0.2.2...v0.2.3) (2020-07-10)


### Bug Fixes

* fix math exporter ([f9db6b8](https://github.com/kanweiwei/easy-editor/commit/f9db6b8331d7d84bc9018166493cb605510fa76a))



## [0.2.2](https://github.com/kanweiwei/easy-editor/compare/v0.2.1...v0.2.2) (2020-07-09)



## [0.2.1](https://github.com/kanweiwei/easy-editor/compare/v0.2.0...v0.2.1) (2020-06-30)


### Bug Fixes

* initValue ([5db5187](https://github.com/kanweiwei/easy-editor/commit/5db51877663913b6f6d1ab810bba1ea069ed9f0c))



# [0.2.0](https://github.com/kanweiwei/easy-editor/compare/v0.1.5...v0.2.0) (2020-06-29)


### Bug Fixes

* control type defination ([2de975c](https://github.com/kanweiwei/easy-editor/commit/2de975c0c68484bf8e80ef7bf3f61e22c23941ac))


### Features

* contentStyle ([20d98b9](https://github.com/kanweiwei/easy-editor/commit/20d98b97bc678e61c4c851144ecba119a38b2eee))



## [0.1.5](https://github.com/kanweiwei/easy-editor/compare/09569762618fbfc660d66fe398cb1764b1ee59a3...v0.1.5) (2020-06-28)


### Bug Fixes

* 🐛 fix image tooltip; publish  v0.1.4 ([36dfe05](https://github.com/kanweiwei/easy-editor/commit/36dfe05be1f09bf51df92cc574a38a8a2a0144a0))
* 🐛 update slate-react ([#13](https://github.com/kanweiwei/easy-editor/issues/13)) ([33f9cd2](https://github.com/kanweiwei/easy-editor/commit/33f9cd25f7bbd37e5a3d85f58164f654843e191f))
* 🐛 类型声明文件找不到 ([780ba4c](https://github.com/kanweiwei/easy-editor/commit/780ba4c1072ba21e20c045f67ade0d1e0499aa90)), closes [#20](https://github.com/kanweiwei/easy-editor/issues/20)
* readOnly时，调用onChange仍起作用 ([c5b96e4](https://github.com/kanweiwei/easy-editor/commit/c5b96e43f0e084cdad2aa001efe9e899e1577a09))


### Features

* 🎸 `value` can  accept Slate Value or Html ([a59670a](https://github.com/kanweiwei/easy-editor/commit/a59670a6ca8bd7c88848fb4df90bb3b7c386b098))
* 🎸 custom control; custom node plugin ([c165252](https://github.com/kanweiwei/easy-editor/commit/c165252ad0aeb45499e25322eadc5aa302c9215f))
* 🎸 init ([0956976](https://github.com/kanweiwei/easy-editor/commit/09569762618fbfc660d66fe398cb1764b1ee59a3))
* 🎸 pdf tmp ([4ffb141](https://github.com/kanweiwei/easy-editor/commit/4ffb1414dda2cf4eee64d2cd1bcc31176b8951b0))
* 🎸 support image、video ([cf86d0c](https://github.com/kanweiwei/easy-editor/commit/cf86d0cb263806dd3fa2dd51847766a79b58968d))
* add changelog ([b54af4a](https://github.com/kanweiwei/easy-editor/commit/b54af4aa16d6382abcdcad6679780d198f7e0d68))
* demo ([8dd2f68](https://github.com/kanweiwei/easy-editor/commit/8dd2f686ea064837ec506b3e61912ff6dcac1698))
* pdf ([4852569](https://github.com/kanweiwei/easy-editor/commit/4852569a088b77dae32b4d7f2bec303a27009d2c))


### Performance Improvements


es/events/paste.js
import _trimInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/trim";
import _forEachInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/for-each";
import _findIndexInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/find-index";
import _mapInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/map";
import _Promise from "@babel/runtime-corejs3/core-js-stable/promise";
import _includesInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/includes";
import { __assign, __awaiter, __generator } from "tslib";
import { Block } from "@zykj/slate";
import { getEventTransfer } from "@zykj/slate-react";
import { get, isEmpty } from "lodash-es";
import { parseFragment, serialize } from "parse5";
import hexToBase64 from "../utils/hexToBase64";
import htmlConvertor, { getAttr } from "../htmlConvertor";
import { getStyleFromString } from "../htmlSerialize"; // eslint-disable-next-line import/extensions

import filterWord from "../utils/filterWord.js";
import getBlobByDataURI from "../utils/getBlobByDataURI";
export default (function (e, change, self, options, beforeUpload) {
  return __awaiter(void 0, void 0, void 0, function () {
    var maxImageHeight, maxImageWidth, imgStyle, transfer, _a, file_1, url, html, reg, localeImgs, supportFileTypes, supportRtfTypes, result, _loop_1, i, rootDom, _getInvalidImgs_, invalidImgs, y, foundIndex, body, document_1, blocks, firstBlock;

    return __generator(this, function (_b) {
      var _context, _context3;

      switch (_b.label) {
        case 0:
          maxImageHeight = get(options, "maxImageHeight");
          maxImageWidth = get(options, "maxImageWidth");
          imgStyle = {};

          if (maxImageHeight) {
            imgStyle.maxHeight = maxImageHeight;
          }

          if (maxImageWidth) {
            imgStyle.maxWidth = maxImageWidth;
          }

          transfer = getEventTransfer(e.nativeEvent);

          if (!transfer) {
            return [2
            /*return*/
            , change];
          }

          _a = transfer.type;

          switch (_a) {
            case "files":
              return [3
              /*break*/
              , 1];

            case "html":
              return [3
              /*break*/
              , 6];

            case "fragment":
              return [3
              /*break*/
              , 12];

            case "node":
              return [3
              /*break*/
              , 12];

            case "rich":
              return [3
              /*break*/
              , 12];

            case "text":
              return [3
              /*break*/
              , 12];
          }

          return [3
          /*break*/
          , 13];

        case 1:
          file_1 = transfer.files[0];
          if (!(file_1 && _includesInstanceProperty(_context = file_1.type).call(_context, "image"))) return [3
          /*break*/
          , 5];
          return [4
          /*yield*/
          , new _Promise(function (resolve, reject) {
            var reader = new FileReader();

            reader.onload = function () {
              resolve(reader.result);
            };

            reader.onerror = function () {
              reject(new Error("error"));
            };

            reader.readAsDataURL(file_1);
          })];

        case 2:
          url = _b.sent();
          if (!(typeof url === "string")) return [3
          /*break*/
          , 5];
          if (!beforeUpload) return [3
          /*break*/
          , 4];
          return [4
          /*yield*/
          , beforeUpload(file_1, url)];

        case 3:
          url = _b.sent();
          _b.label = 4;

        case 4:
          change = change.insertInline({
            object: "inline",
            type: "image",
            isVoid: true,
            data: {
              style: imgStyle,
              src: url
            }
          });
          self.onChange(change);
          _b.label = 5;

        case 5:
          return [2
          /*return*/
          , false];

        case 6:
          html = filterWord(transfer.html);
          reg = /<(?=span|font|\/span|\/font).*?>/g;
          html = html.replace(reg, "");
          html = html.replace(/[．|．]/g, ".");
          if (!/file:[\s\S]*.\.(png|jpg|jpeg)/.test(html)) return [3
          /*break*/
          , 11];
          localeImgs = [];
          supportFileTypes = [{
            rtfType: "pngblip",
            imgType: "png",
            fileType: "image/png"
          }, {
            rtfType: "jpegblip",
            imgType: "jpg",
            fileType: "image/jpg"
          }, {
            rtfType: "jpgblip",
            imgType: "jpg",
            fileType: "image/jpg"
          }];
          supportRtfTypes = _mapInstanceProperty(supportFileTypes).call(supportFileTypes, function (n) {
            return n.rtfType;
          });
          reg = new RegExp("(" + supportRtfTypes.join("|") + ")[^}]*}([^}]*)}", "gim");
          result = transfer.rich.match(reg);
          if (!(result && result.length > 0)) return [3
          /*break*/
          , 10];

          _loop_1 = function _loop_1(i) {
            var m, _a, str, type, hexData, fileType, typeIndex, base64Data, dataURI, url;

            return __generator(this, function (_b) {
              switch (_b.label) {
                case 0:
                  m = result[i].replace(/\s/gm, "");
                  reg = new RegExp("(" + supportRtfTypes.join("|") + ")[^}]*}([^}]*)}", "i");
                  _a = m.match(reg), str = _a[0], type = _a[1], hexData = _a[2];
                  fileType = "image/jpg";
                  typeIndex = _findIndexInstanceProperty(supportFileTypes).call(supportFileTypes, function (n) {
                    return n.rtfType === type;
                  });

                  if (typeIndex > -1) {
                    // imgType = supportFileTypes[typeIndex].imgType;
                    fileType = supportFileTypes[typeIndex].fileType;
                  }

                  base64Data = hexToBase64(hexData);
                  dataURI = "data:" + fileType + ";base64," + base64Data;
                  url = dataURI;
                  if (!beforeUpload) return [3
                  /*break*/
                  , 2];
                  return [4
                  /*yield*/
                  , beforeUpload(getBlobByDataURI(dataURI, fileType), dataURI)];

                case 1:
                  url = _b.sent();
                  _b.label = 2;

                case 2:
                  localeImgs.push(url);
                  return [2
                  /*return*/
                  ];
              }
            });
          };

          i = 0;
          _b.label = 7;

        case 7:
          if (!(i < result.length)) return [3
          /*break*/
          , 10];
          return [5
          /*yield**/
          , _loop_1(i)];

        case 8:
          _b.sent();

          _b.label = 9;

        case 9:
          i++;
          return [3
          /*break*/
          , 7];

        case 10:
          rootDom = parseFragment(html);

          _getInvalidImgs_ = function getInvalidImgs_1(nodes, arr) {
            _forEachInstanceProperty(nodes).call(nodes, function (node) {
              if (node.childNodes && node.childNodes.length > 0) {
                _getInvalidImgs_(node.childNodes, arr);
              }

              if (node.nodeName.toLowerCase() === "img") {
                var src = getAttr(node.attrs, "src");

                if (/file:[\s\S]*.\.(png|jpg|jpeg)/.test(src)) {
                  arr.push(node);
                }
              }
            });
          };

          invalidImgs = [];

          _getInvalidImgs_(rootDom.childNodes, invalidImgs);

          if (invalidImgs.length > 0) {
            for (y = 0; y < invalidImgs.length; y++) {
              var _context2;

              foundIndex = _findIndexInstanceProperty(_context2 = invalidImgs[y].attrs).call(_context2, function (n) {
                return n.name === "src";
              });

              if (foundIndex > -1 && localeImgs[y]) {
                invalidImgs[y].attrs[foundIndex].value = localeImgs[y];
              }
            }
          }

          html = serialize(rootDom);
          body = window.document.createElement("body");
          body.innerHTML = html;

          if (_trimInstanceProperty(_context3 = body.innerText).call(_context3).length === 0 && invalidImgs.length > 0) {
            _forEachInstanceProperty(invalidImgs).call(invalidImgs, function (n) {
              var width = getAttr(n.attrs, "width");
              var height = getAttr(n.attrs, "height");
              var style = "";

              if (width) {
                style += "width: " + (Number(width) ? width + "px" : width) + ";";
              }

              if (height) {
                style += "height: " + (Number(height) ? height + "px" : height) + ";";
              }

              var data = {};
              data.src = getAttr(n.attrs, "src");

              if (style) {
                data.style = getStyleFromString(style);
              }

              if (!isEmpty(imgStyle)) {
                var maxWidth = width / height * maxImageHeight;
                data.style = __assign(__assign(__assign({}, data.style), imgStyle), {
                  maxWidth: maxWidth
                });
              }

              change.insertInline({
                object: "inline",
                type: "image",
                isVoid: true,
                data: data
              });
            });

            return [2
            /*return*/
            , self.onChange(change)];
          }

          change = change.insertFragment(htmlConvertor.deserialize(html).document);
          return [2
          /*return*/
          , self.onChange(change)];

        case 11:
          document_1 = htmlConvertor.deserialize(html).document;
          blocks = change.value.blocks;
          firstBlock = blocks.first();

          if (firstBlock.text.length > 0 || firstBlock.findDescendant(function (n) {
            return n.type === "image";
          })) {
            change = change.insertBlock(Block.fromJSON({
              type: "div",
              object: "block",
              nodes: []
            }));
          }

          change = change.insertFragment(document_1);
          return [3
          /*break*/
          , 14];

        case 12:
          if (transfer.text) {
            try {
              change = change.insertText(transfer.text);
            } catch (err) {
              console.error(err.message);
            }
          } // return false;


          return [3
          /*break*/
          , 14];

        case 13:
          if (transfer.text) {
            try {
              change = change.insertText(transfer.text);
            } catch (err) {
              console.error(err.message);
            }
          }

          _b.label = 14;

        case 14:
          e.preventDefault();

          if (transfer.type !== "files") {
            self.onChange(change);
          }

          return [2
          /*return*/
          , false];
      }
    });
  });
});